quickwit-actors/src/actor.rs

```diff
@@ -0,0 +1,287 @@
+    actor_state: AtomicState,
+}
+
+impl<A: Actor> ActorContext<A> {
+    pub(crate) fn new(
+        self_mailbox: Mailbox<A::Message>,
+        kill_switch: KillSwitch,
+        scheduler_mailbox: Mailbox<SchedulerMessage>,
+    ) -> Self {
+        let actor_instance_name = self_mailbox.actor_instance_name();
+        ActorContext {
+            inner: ActorContextInner {
+                actor_instance_name,
+                self_mailbox,
+                progress: Progress::default(),
+                kill_switch,
+                scheduler_mailbox,
+                actor_state: AtomicState::default(),
+            }
+            .into(),
+        }
+    }
+
+    pub fn mailbox(&self) -> &Mailbox<A::Message> {
+        &self.self_mailbox
+    }
+
+    pub fn actor_instance_name(&self) -> &str {
+        &self.actor_instance_name
+    }
+
+    /// This function returns a guard that prevents any supervisor from identifying the
+    /// actor has dead.
+    /// The protection last as long as the `ProtectZoneGuard` it returns.
+    ///
+    /// In an ideal world, you should never need to call this function.
+    /// It is only useful in some corner like, like a calling a long blocking
+    /// from an external library that you trust.
+    pub fn protect_zone(&self) -> ProtectZoneGuard {
+        self.progress.protect_zone()
+    }
+
+    /// Get a copy of the actor kill switch.
+    /// This should rarely be used.
+    /// For instance, when quitting from the process_message function, prefer simply
+    /// returning `Error(ActorTermination::Failure(..))`
+    pub fn kill_switch(&self) -> &KillSwitch {
+        &self.kill_switch
+    }
+
+    pub(crate) fn progress(&self) -> &Progress {
+        &self.progress
+    }
+    /// Record some progress.
+    /// This function is only useful when implementing actors that may takes more than
+    /// `HEARTBEAT` to process a single message.
+    /// In that case, you can call this function in the middle of the process_message method
+    /// to prevent the actor from being identified as blocked or dead.
+    pub fn record_progress(&self) {
+        self.progress.record_progress();
+    }
+
+    pub(crate) fn get_state(&self) -> ActorState {
+        self.actor_state.get_state()
+    }
+
+    pub(crate) fn pause(&mut self) {
+        self.actor_state.pause();
+    }
+
+    pub(crate) fn resume(&mut self) {
+        self.actor_state.resume();
+    }
+
+    pub(crate) fn terminate(&mut self, termination: &ActorTermination) {
+        if termination.is_failure() {
+            self.kill_switch().kill();
+        }
+        self.actor_state.terminate();
+    }
+}
+
+impl<A: SyncActor> ActorContext<A> {
+    /// Sends a message to the actor being the mailbox.
+    ///
+    /// This method hides logic to prevent an actor from being identified
+    /// as frozen if the destination actor channel is saturated, and we
+    /// are simply experiencing back pressure.
```

Comment:
We need to make sure to avoid cycles, a pipeline should only always go forward

Reply:
That's a famous problem associated with actors in general yes.